src/passes/StackIR.cpp

```diff
@@ -0,0 +1,393 @@
+
+  bool modifiesBinaryenIR() override { return false; }
+
+  void doWalkFunction(Function* func) {
+    std::cout << func->name << ":\n";
+    if (func->stackIR) {
+      std::cout << *func->stackIR;
+    } else {
+      std::cout << " (no stack ir)";
+    }
+    std::cout << '\n';
+  }
+};
+
+Pass* createPrintStackIRPass() {
+  return new PrintStackIR();
+}
+
+// Optimize
+
+class StackIROptimizer {
+  Function* func;
+  PassOptions& passOptions;
+  StackIR& insts;
+
+public:
+  StackIROptimizer(Function* func, PassOptions& passOptions) :
+    func(func), passOptions(passOptions), insts(*func->stackIR.get()) {
+    assert(func->stackIR);
+  }
+
+  void run() {
+    dce();
+    // FIXME: local2Stack is currently rather slow (due to localGraph),
+    //        so for now run it only when really optimizing
+    if (passOptions.optimizeLevel >= 3 || passOptions.shrinkLevel >= 1) {
+      local2Stack();
+    }
+    removeUnneededBlocks();
+    dce();
+  }
+
+private:
+  // Passes.
+
+  // Remove unreachable code.
+  void dce() {
+    bool inUnreachableCode = false;
+    for (Index i = 0; i < insts.size(); i++) {
+      auto* inst = insts[i];
+      if (!inst) continue;
+      if (inUnreachableCode) {
+        // Does the unreachable code end here?
+        if (isControlFlowBarrier(inst)) {
+          inUnreachableCode = false;
+        } else {
+          // We can remove this.
+          removeAt(i);
+        }
+      } else if (inst->type == unreachable) {
+        inUnreachableCode = true;
+      }
+    }
+  }
+
+  // If ordered properly, we can avoid a set_local/get_local pair,
+  // and use the value directly from the stack, for example
+  //    [..produce a value on the stack..]
+  //    set_local $x
+  //    [..much code..]
+  //    get_local $x
+  //    call $foo ;; use the value, foo(value)
+  // As long as the code in between does not modify $x, and has
+  // no control flow branching out, we can remove both the set
+  // and the get.
+  void local2Stack() {
+    // We use the localGraph to tell us if a get-set pair is indeed
+    // a set that is read by that get, and only that get. Note that we run
+    // this on the Binaryen IR, so we are assuming that no previous opt
+    // has changed the interaction of local operations.
+    // TODO: we can do this a lot faster, as we just care about linear
+    //       control flow.
+    LocalGraph localGraph(func);
+    localGraph.computeInfluences();
+    // We maintain a stack of relevant values. This contains:
+    //  * a null for each actual value that the value stack would have
+    //  * an index of each SetLocal that *could* be on the value
+    //    stack at that location.
+    const Index null = -1;
+    std::vector<Index> values;
+    // We also maintain a stack of values vectors for control flow,
+    // saving the stack as we enter and restoring it when we exit.
+    std::vector<std::vector<Index>> savedValues;
+#ifdef STACK_OPT_DEBUG
```

Comment:
Should this be a `#define`? Probably for now I guess, because we won't want to use this unless developing the stack optimizer. Can change it later if we need to, for now probably best to keep it simple.